lib/smtp.c

```diff
@@ -517,10 +517,25 @@ static CURLcode smtp_perform_mail(struct connectdata *conn)
   if(!data->set.str[STRING_MAIL_FROM])
     /* Null reverse-path, RFC-5321, sect. 3.6.3 */
     from = strdup("<>");
-  else if(data->set.str[STRING_MAIL_FROM][0] == '<')
-    from = aprintf("%s", data->set.str[STRING_MAIL_FROM]);
-  else
-    from = aprintf("<%s>", data->set.str[STRING_MAIL_FROM]);
+  else {
+    struct hostname mail_from = { NULL, NULL, NULL, NULL };
+
+    mail_from.name = data->set.str[STRING_MAIL_FROM];
+
+    /* Convert the MAIL FROM address to IDN (if necessary) */
+    if(conn->proto.smtpc.utf8_supported) {
```

Comment:
It doesn't make sense to check for UTF8 for IDN conversion:
1. The input IDN name is not UFT-8
2. The output puny code is not UTF-8 either